Add concurrency and experimental tagging to CI matrix

NSeydoux opened the pull request:
- **Experimental tags in e2e test matrices**: If a config is tagged `experimental` in the test matrix, it failing will not interrupt other jobs of the matrix. https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#expanding-or-adding-matrix-configurations explains how the `include` works in matrices strategy. Namely, in our case `experimental` being set to false by default, all the environments defined in the non-`include` matrix will run with the `experimental: false` tag, and an additional test matrix with `{ environment-name: "ESS Dev-Next", experimental: true}` will be added. The latter will have the `continue-on-error` tag, which means it failing will not interrupt the other e2e jobs.
- **Concurrency groups**: See https://docs.github.com/en/actions/using-jobs/using-concurrency#example-only-cancel-in-progress-jobs-or-runs-for-the-current-workflow. Adding a job to a concurrency group interrups other jobs running from the same group, instead of having the running jobs complete first.

NSeydoux committed:
Enable experimental runs

If a job is tagged `experimental` in the test matrix, it failing will not interrupt other jobs of the matrix

NSeydoux committed:
Add concurrency config

The concurrency config interrups jobs running when new jobs are added to the queue, instead of having the running jobs complete first.

NSeydoux committed:
Lint

NSeydoux committed:
fixup! Enable experimental runs

NSeydoux commented:
Note to reviewer: a `Dev-Next` browser-based test just failed while the `PodSpaces` one was running, not cancelling it, so the config looks about right.

ThisIsMissEm commented on .github/workflows/audit.yml:
```diff
@@ -4,7 +4,9 @@ on:
     - cron: "40 10 * * *"
-
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
```
Just to double check `github.ref` is the pull request ID?

NSeydoux replied:
That's the branch: https://docs.github.com/en/actions/learn-github-actions/environment-variables#default-environment-variables

ThisIsMissEm commented on .github/workflows/e2e-browser.yml:
```diff
@@ -14,9 +16,14 @@ jobs:
     strategy:
       matrix:
         environment-name: ["ESS PodSpaces"]
+        experimental: [false]
```
```suggestion
        experimental: false
```

i don't think you need the brackets here

NSeydoux replied:
That's also true for the environment name actually

ThisIsMissEm replied:
so `environmentName` should have it, as there should be multiple there `environmentName: [ "ESS PodSpaces", "ESS Dev-Next" ]`

ThisIsMissEm commented on .github/workflows/e2e-browser.yml:
```diff
@@ -14,9 +16,14 @@ jobs:
+    continue-on-error: ${{ matrix.experimental }}
     strategy:
       matrix:
         environment-name: ["ESS PodSpaces"]
```
```suggestion
        environment-name: ["ESS PodSpaces", "ESS Dev-Next"]
```

this is how I've seen it done elsewhere

NSeydoux replied:
I'll give it a try, but I don't think this will work, because this will add a `{ env: "Dev-Next", experimental: "false" }` in the run matrix, which we don't want. The `include` block would just add an additional `{ env: "Dev-Next", experimental: "true" }` in any case, since the `experimental: true` flag would override the value in the matrices otherwise, which is always defined and `true`.

NSeydoux replied:
Yup confirmed, the correct output there is obtained not including `ESS Dev-Next` in the canonical matrix, but only in the `include` block. https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#expanding-or-adding-matrix-configurations provides helpful context. 
![correct](https://user-images.githubusercontent.com/8236166/202694321-b9b81881-6b35-4364-94af-c9e1fa54ecf2.png)
![incorrect](https://user-images.githubusercontent.com/8236166/202694324-53acebd7-91b5-4859-b14a-9f873241e549.png)

ThisIsMissEm commented on .github/workflows/release.yml:
```diff
@@ -4,7 +4,9 @@ on:
-
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
```
This probably shouldn't have this in place, as all release jobs must be completed entirely

NSeydoux replied:
86b245eae7aab630099765a786894628cf3a8a09

ThisIsMissEm reviewed (changes requested):
Requesting changes on the release config

NSeydoux committed:
Do not interrupt release workflows

NSeydoux committed:
Fix expermiental matrices

`include` variables don't have to be an array

NSeydoux committed:
Add dev-next to canonical envs

NSeydoux committed:
Revert "Add dev-next to canonical envs"

This reverts commit e991dfc23311b1c207a993c9dda3f682bac27f5a.

NSeydoux committed:
Fix playwright upload path

NSeydoux committed:
Fix experimental matrix e2e-node

NSeydoux committed:
Fix env secret names

NSeydoux committed:
Disable e2e tests for Dev-Next

NSeydoux committed:
Update shared action repo name

ThisIsMissEm commented on .github/workflows/e2e-node.yml:
```diff
@@ -15,11 +17,19 @@ jobs:
     strategy:
       matrix:
         os: [ubuntu-latest]
         node-version: [16.x]
```
our version range says we still support node.js 14.x, so we should test on that too: https://github.com/inrupt/solid-client-notifications-js/blob/main/package.json#L98
```suggestion
        node-version: [16.x, 14.x]
```

ThisIsMissEm reviewed (changes requested):
Looking good, one minor issue

NSeydoux committed:
Update .github/workflows/e2e-node.yml

Co-authored-by: Emelia Smith <ThisIsMissEm@users.noreply.github.com>

NSeydoux committed:
Align with authn feedback

ThisIsMissEm reviewed (changes requested):
Can you also rename `e2e/node/setup.ts` to `e2e/node/jest.setup.ts`, and merge `jest.config.ts` and `jest.e2e.config.ts` like we've done in the other repositories? Also, we can delete `.testcaferc`

NSeydoux committed:
Rename e2e setup file

NSeydoux committed:
Install latest jest

NSeydoux committed:
Convert jest config to TS

NSeydoux committed:
fixup! Install latest jest

NSeydoux committed:
Migrate to lates jest

Disable testing the dynamic loading of solid-client-authn-browser because of jose issues as ESM in jsdom

NSeydoux committed:
Remove legacy file

NSeydoux commented:
Actually this project hadn't been migrated to the latest version of jest, which is now done.
